crypto/aria/aria.c

```diff
@@ -1,454 +1,1289 @@
- */
-int aria_set_decrypt_key(const unsigned char *userKey, const int bits,
-                         ARIA_KEY *key)
-{
-    ARIA_KEY ek;
-    const int r = aria_set_encrypt_key(userKey, bits, &ek);
-    unsigned int i, rounds = ek.rounds;
-
-    if (r == 0) {
-        key->rounds = rounds;
-        memcpy(key->rd_key[0], ek.rd_key[rounds], sizeof(key->rd_key[0]));
-        for (i = 1; i < rounds; i++)
-            a(key->rd_key[i], ek.rd_key[rounds - i]);
-        memcpy(key->rd_key[rounds], ek.rd_key[0], sizeof(key->rd_key[rounds]));
-    }
-    return r;
-}
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* 
+ * Copyright (C) 2017 National Security Research Institute. All Rights Reserved.
+ *
+ * Information for ARIA
+ *     http://210.104.33.10/ARIA/index-e.html (English)
+ *     http://seed.kisa.or.kr/ (Korean)
+ *
+ * Public domain version is distributed above.
+ */
+
+/* ====================================================================
+ * Copyright (c) 2017 Oracle and/or its affiliates.  All rights reserved.
+ */
+
+#include <openssl/e_os2.h>
+#include "internal/aria.h"
+
+#include <assert.h>
+#include <string.h>
+#include <stddef.h>
+
+#ifndef OPENSSL_SMALL_FOOTPRINT
+
+/* Begin macro */
+#ifndef __has_builtin
+# define __has_builtin(x) 0
+#endif
+
+#if !defined(B_ENDIAN) && !defined(L_ENDIAN)
+# if defined(__BYTE_ORDER__) && defined(__ORDER_BIG_ENDIAN__) && \
+    (__BYTE_ORDER__ == __ORDER_BIG_ENDIAN__)
+#  define B_ENDIAN
+# elif defined(__BIG_ENDIAN__) || \
+    defined(_M_PPC) || defined(_M_MPPC) || \
+    defined(__ARMEB__) || defined(__THUMBEB__) || defined(__AARCH64EB__) || \
+    defined(_MIPSEB) || defined(__MIPSEB) || defined(__MIPSEB__)
+#  define B_ENDIAN
+# endif
+#endif
+
+/* rotation */
+#if defined(_MSC_VER)
+# include <stdlib.h>
+# define rotl32(v, r) _lrotl(v, r)
+# define rotr32(v, r) _lrotr(v, r)
+#else
+# define rotl32(v, r) (((uint32_t)(v) << (r)) | ((uint32_t)(v) >> (32 - r)))
+# define rotr32(v, r) (((uint32_t)(v) >> (r)) | ((uint32_t)(v) << (32 - r)))
+#endif
+
+#if defined(_MSC_VER) && _MSC_VER >= 1400
+# define bswap32(v) _byteswap_ulong(v)
+#elif __has_builtin(__builtin_bswap32)
+# define bswap32(v) __builtin_bswap32(v)
+#elif (defined(__GNUC__) && \
+    ((__GNUC__ > 4) || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3)))
+# define bswap32(v) __builtin_bswap32(v)
+#else
+# define bswap32(v)                                         \
+    (((v) << 24) ^ ((v) >> 24) ^                            \
+    (((v) & 0x0000ff00) << 8) ^ (((v) & 0x00ff0000) >> 8)
+#endif
+
+/* Load Big-Endian Integer */
+#ifndef B_ENDIAN
+# define loadBE32(v) bswap32(v)
+#else
+# define loadBE32(v) ((uint32_t)(v))

```

Comment:
What about creating if-statement**s** for already well-known architecture, like the Camellia's code?
If we can assure align problem, we can get a significant speed improvement.